Update scripting.md to update details and fix inconsistencies (#22464)
OpenRCT2/OpenRCT2: distribution/scripting.md

- Mac: `/Users/YourName/Library/Application Support/OpenRCT2`
- Linux: `$XDG_CONFIG_HOME/OpenRCT2` or in its absence `$HOME/.config/OpenRCT2`

> You can also find the user directory via the "Open custom content folder" option under the red toolbox button in the game's title screen.

Reviewer: ```suggestion
You can also find the user directory via the “Open custom content folder” option under the red toolbox button in the game's title screen.
```

OpenRCT2 will load every single file with the extension `.js` in this directory recursively. So if you want to prevent a plug-in from being used, you must move it outside this directory, or rename it so the filename does not end with `.js`.

## Script types

There are three types of scripts:
* Local
* Remote
* Intransient

> Local scripts can **not** alter the game state directly. This allows each player to enable any local script for their own game without other players in a multiplayer server also needing to also enable the same script. Local scripts can interact with the game by mimicing a player's actions through the built-in "game actions". These scripts tend to provide extra tools for productivity, or new windows containing information.

Reviewer: ```suggestion
Local scripts **cannot** alter the game state directly. This allows each player to enable any local script for their own game without other players in a multiplayer server also needing to also enable the same script. Local scripts **can** interact with the game by mimicking a player's actions through the built-in “game actions”. These scripts tend to provide extra tools for productivity, or new windows containing information.
```

> Local scripts **cannot** alter the game state directly. This allows each player to enable any local script for their own game without other players in a multiplayer server also needing to also enable the same script. Local scripts **can** interact with the game by mimicking a player's actions through the built-in “game actions”. These scripts tend to provide extra tools for productivity, or new windows containing information.

Reviewer 2: The section "What does the error 'Game state is not mutable in this context' mean?" is similarly wrong.

> Local scripts **cannot** alter the game state directly. This allows each player to enable any local script for their own game without other players in a multiplayer server also needing to also enable the same script. Local scripts **can** interact with the game by mimicking a player's actions through the built-in “game actions”. These scripts tend to provide extra tools for productivity, or new windows containing information.

Author: Yes, I can clarify the difference between singleplayer and multiplayer as you mentioned, as local vs. remote has no differences there. 

For multiplayer I do think "remote" by default is better. It's better that the plugin you bring doesn't load at all than that it loads but breaks/errors because the plugin developer didn't take mutable context in account. Having it as "remote" by default is more intuitive for users and developers to understand why it doesn't load in multiplayer if multiplayer hasn't been accounted for, as it's "oh I need to do something extra for multiplayer" (as you explained).

Note that remove server plugins can exist without custom game actions; e.g. UI popups/welcome messages when a player joins, spam protection, tools to kick/ban players, block certain built-in game actions by subscribing to the hooks, run certain other code tied to hooks ("interval.day", "ratings.calculate" etc.).

> Local scripts **cannot** alter the game state directly. This allows each player to enable any local script for their own game without other players in a multiplayer server also needing to also enable the same script. Local scripts **can** interact with the game by mimicking a player's actions through the built-in “game actions”. These scripts tend to provide extra tools for productivity, or new windows containing information.

Reviewer 2: I think the main point of the local vs remote question is "What does my plugin want to do?" and not "Should my plugin target singleplayer or multiplayer?" and I think in the past there was a lot of confusion about this, maybe because the naming could suggest that "local" is for single player and "remote" is for multiplayer, which is absolutely not the case.

    main: main
});
```

This will log a message to the terminal screen (`stdout`) when you open any park. If you are on Windows, make sure to run `openrct2.com` instead of `openrct2.exe` so you can interact with the `stdin` / `stdout` console. The console is a JavaScript interpreter (REPL), this means you can write and test expressions similar to the console found in web browsers when you press `F12`. When you make changes to your script, you must exit your current game and open it again for the script to reload... unless you use the hot reload feature.

### Script types

There are three types of scripts that can influence how the game loads the script:

* Local
* Remote
* Intransient

#### Singleplayer

In singleplayer all plugin types are unrestricted. Local and remote plugins reload each time a scenario or a save file is opened. Plugins of both types can alter game state directly or execute either built-in game actions or custom game actions.

Intransient plugins load when the game starts and stays active until the game is shutdown.

#### Multiplayer

In multiplayer local and remote plugins are different in when they are loaded and how they can alter the game state.

> **Local plugins** will load on any client in multiplayer, which will allow each player to bring any local script for their own game without other players in a multiplayer server also needing to also enable the same script.

Author: What is the misleading part? Also not sure what "as usual" refers too? Singleplayer? That may be misleading because local plugins are not completely "as usual" as singleplayer.

I'll fix the double "also" though, bad copy paste mistake. 🙂

> **Local plugins** will load on any client in multiplayer, which will allow each player to bring any local script for their own game without other players in a multiplayer server also needing to also enable the same script.

Reviewer 2: "Local plugins will load on any client in multiplayer" can be interpreted as any client that joins will load your local script.
Yes, with usual I meant compared to singleplayer.

> **Local plugins** will load on any client in multiplayer, which will allow each player to bring any local script for their own game without other players in a multiplayer server also needing to also enable the same script.

Author: Good point. Added a few extra words that will hopefully clarify that confusion. 🙂

- Intransient plugins could provide functionality across different screens such as the title screen, and can be used for plugins like the title sequence editor or global dashboards.

### Plugin licence

The authors must also define a licence for the plugin, making it clear to the community whether that plugin can be altered, copied, etc. A good reference material is listed on [ChooseALlicense](https://choosealicense.com/appendix/), try to pick one of them and use its corresponding identifier, as listed on [SPDX](https://spdx.org/licenses/).

### Hot reload

The hot reload feature can be enabled by editing your `config.ini` file and setting `enable_hot_reloading` to `true` under `[plugin]`. When this is enabled, the game will auto-reload the script in real-time whenever you save your JavaScript file. This allows rapid development of plugins as you can write code and quickly preview your changes, such as closing and opening a specific custom window on startup. A demonstration of this can be found on YouTube: [OpenRCT2 plugin hot-reload demo](https://www.youtube.com/watch?v=jmjWzEhmDjk)

## Game actions

> Game actions are actions that players can invoke in games. Here is an example flow of a game action such as opening the park:

Reviewer 2: Please add some intuition how to use actions (especially compared to direct data access):
- for everything that the player can do to change the game state, there is a game action
- plugins can execute any of these actions as if they were a player (it is actually never mentioned here)
- **using actions is the prefered way to change the game state**, since it automatically ensures multiplayer-safety and sanity checks (cash, clearance, ...)
- but, if the plugin really needs to use direct game state changes, it is strongly advised to make the plugin remote and wrap all game state changes in custom actions. Otherwise, the plugin will not work in multiplayer and likely cause disconnects. Even if you as the plugin developer do not want to specifically target multiplayer, it is still better to follow these guidelines to prevent disconnects of your users.

Please copy or modify as you see fit.
As a side note, if it wouldn't absolutely destroy backwards compatibility, I would really like to disable all direct game state changes outside of custom actions.

In multiplayer, there are some additional steps after step 3:

4. If the query succeeds and the game is multiplayer: game action is send to the server.
5. Server checks if the player has permissions to perform the game action.
6. Server calls the query method again.
7. If the permission check or the query fails, the server sends an error message back to the client.
8. If the perission check and query succeed: send action to all players to execute action at a specified tick.
9. All player clients receive the game action and execute it at the specified tick.
10. A tree is placed at the specified location on all clients that are connected to the server.

This sequence of steps ensures that every player client executes the game action exactly in the same way on the exact same game tick (synchronized execution). Furthermore, it allows the server to validate that the action is allowed and does not fail due to permission or another player executing a conflicting action just before it. This is why there is a noticeable delay when constructing in multiplayer games, as the client has to wait for the server to acknowledge the action and reply with the tick number to execute it on.

> Plugins can also use game actions to alter the game in the same way players can. Game actions are the **recommended approach** to mutating game state, as it automatically ensures multiplayer-safety and various sanity checks for money, clearance checks, and others.

Author: Fixed, thanks! 🙂

> What are the limits?

Scripts can consist of any behaviour and have a large memory pool available to them. The speed will vary depending on the hardware and system executing them. The scripts are interpreted, so do not expect anywhere close to the performance of native code. In most scenarios this should be satisfactory, but a random map generator, or genetic algorithm for building roller coasters might struggle. Like any language, there will be tricks to optimising JavaScript and the use of the OpenRCT2 APIs. [Duktape also provides some engine-specific performance tips](https://wiki.duktape.org/performance). 

> The APIs for OpenRCT2 try to provide access to the game's data structures as much as possible but we can only add so many at a time. The best way to grow the plug-in system is to add APIs on-demand. So if you find an API is missing, please raise an issue for it on GitHub and also feel free to discuss it on our Discord and to submit a pull request afterwards.

Reviewer: ```suggestion
The APIs for OpenRCT2 try to provide access to the game’s data structures as much as possible but we can only add so many at a time. The best way to grow the plug-in system is to add APIs on-demand. So if you find an API is missing, please raise an issue for it on GitHub and also feel free to discuss it on our Discord and to submit a pull request afterwards.
```

> What is ```targetApiVersion```

> In case there are breaking Api changes plugins can use this to keep the old API behavior. For example in version 34 ```Entity.type``` would no longer return "peep" for guests and staff, instead it would return either "guest" or "staff", so if your plugin expects "peep" you can specify the version 33 to keep the old behavior. See the list of breaking changes. If this is not specified it will default to version 33 and the plugin will log an error on startup, it is recommended to specify the current api version.

Reviewer: ```suggestion
In case there are breaking API changes, plugins can use this to keep the old API behaviour. For example in version 34 ```Entity.type``` would no longer return `peep` for guests and staff, instead it would return either `guest` or `staff`, so if your plugin expects `peep` you can specify the version 33 to keep the old behaviour. See the list of breaking changes. If this is not specified it will default to version 33 and the plugin will log an error on startup, it is recommended to specify the current API version.
```

Also: this refer to a list of breaking changes, but does not specify where.

});
```

> Other hooks include receiving a chat message in multiplayer, or a ride's ratings have been recalculated.

Reviewer: ```suggestion
Other hooks include receiving a chat message in multiplayer, or recalculation of a ride’s ratings.
```

This sequence of actions ensures that every player execute the action exactly in the same way on the exact same game tick. It also allows the server to validate that the action is allowed and does not fail due to permission or another player executing a conflicting action just before it. This is why there is a noticeable delay when constructing in multiplayer games, as the client has to wait for the server to acknowledge the action and reply with the tick number to execute it on, since we do not yet have any rollback support.

> Besides built-in game actions, remote scripts can also register their own game actions ("custom game actions"). These custom actions allow custom game state mutations to be executed in sync for all players on the server. Beware that the plugin will also be responsible for proper permission checks, as custom game actions have no permission checks by default.

Reviewer: ```suggestion
Besides built-in game actions, remote scripts can also register their own game actions (“custom game actions”). These custom actions allow custom game state mutations to be executed in sync for all players on the server. Beware that the plugin will also be responsible for proper permission checks, as custom game actions have no permission checks by default.
```

---

> What is `targetApiVersion`?

In case there are breaking API changes, plugins can use this to keep the old API behaviour. For example in version 34 `Entity.type` would no longer return `peep` for guests and staff, instead it would return either `guest` or `staff`, so if your plugin expects `peep` you can specify the version 33 to keep the old behaviour. See the [list of breaking changes](#breaking-changes). If this is not specified it will default to version 33 and the plugin will log an error on startup, it is recommended to specify the current API version.

---

> How do I prevent my script from running on older versions of OpenRCT2 that do not support all the APIs I require?

When registering your plugin, you can specify a minimum required API version like below:

```js
registerPlugin({
    name: 'Your Plugin',
    version: '1.2', // Your plugin version
    authors: ['Your Name'],

>     type: 'remote',

Reviewer: ```suggestion
    type: 'remote',
    licence: 'MIT', // Specify your licence here (MIT is just a recommendation). You can use the `SPDX` identifier, or point to a file that contains the licence text.
```

    minApiVersion: 7, // OpenRCT2 plugin API v7 or higher is required to run your plugin

>     main: main

Reviewer: Plugins should also specify their licence. (We generally recommend `MIT` as the default licence here.)

---

> What is `targetApiVersion`?

In case there are breaking API changes, plugins can use this to keep the old API behaviour. For example in version 34 `Entity.type` would no longer return `peep` for guests and staff, instead it would return either `guest` or `staff`, so if your plugin expects `peep` you can specify the version 33 to keep the old behaviour. See the [list of breaking changes](#breaking-changes). If this is not specified it will default to version 33 and the plugin will log an error on startup, it is recommended to specify the current API version.

---

> How do I prevent my script from running on older versions of OpenRCT2 that do not support all the APIs I require?

When registering your plugin, you can specify a minimum required API version like below:

```js
registerPlugin({
    name: 'Your Plugin',
    version: '1.2', // Your plugin version
    authors: ['Your Name'],
    type: 'remote',

>     licence: 'MIT', // Specify your licence here (MIT is just a recommendation). You can use the `SPDX` identifier, or point to a file that contains the licence text.

Reviewer: ```suggestion
    licence: 'MIT',
```

This sequence of actions ensures that every player execute the action exactly in the same way on the exact same game tick. It also allows the server to validate that the action is allowed and does not fail due to permission or another player executing a conflicting action just before it. This is why there is a noticeable delay when constructing in multiplayer games, as the client has to wait for the server to acknowledge the action and reply with the tick number to execute it on, since we do not yet have any rollback support.

> Besides built-in game actions, remote scripts can also register their own game actions (“custom game actions”). These custom actions allow custom game state mutations to be executed in sync for all players on the server. Beware that the plugin will also be responsible for proper permission checks, as custom game actions have no permission checks by default.

Reviewer 2: Technically, any plugin can define and use custom actions, not just remote plugins.

Author: Technically yes, but practically there's not much benefit. In multiplayer clients who don't have the local plugin won't execute the custom action.
